pennylane/argmap.py

```diff
@@ -0,0 +1,251 @@
+    ``int`` instead, indexing a one-dimensional array.
+
+    If the argument index (first entry) of a key is ``None``, this indicates that there is
+    only one argument in the QNode, and all keys of the ``ArgMap`` should have the form
+    ``(None, tuple[int] or int)``.
+    If the parameter index (second entry) of a key is ``None``, the respective argument
+    is a scalar and no other key with the same argument index can exist in the ``ArgMap``.
+    If both entries are ``None``, i.e. ``key=(None, None)``, the QNode only takes a single
+    scalar argument.
+
+    The main feature of an ``ArgMap`` is its flexible access: Users can specify only an
+    argument index or only a parameter index if this fully uniquely describes the
+    parameter.
+
+    **Example**
+    Suppose we have the following ``QNode`` instance:
+
+    .. code-block:: python
+
+        dev = qml.device('default.qubit', wires=2)
+        @qml.qnode(dev)
+        def circuit(arg1, arg2, arg3):
+            qml.Rot(arg1, wires=0)
+            qml.RX(arg2, wires=0)
+            qml.Rot(arg3, wires=1)
+            return qml.expval(qml.PauliY(0) @ qml.PauliY(1))
+
+    Then we might want to set up an ``ArgMap`` to store the generators of the rotations in
+    this circuit, depending on the parameter that controls them:
+
+    .. code-block:: python
+
+        generators = qml.ArgMap({
+            (0, 0): qml.PauliZ(0),
+            (0, 1): qml.PauliY(0),
+            (0, 2): qml.PauliZ(0),
+            1: qml.PauliX(0),
+            (1, 0): qml.PauliZ(1),
+            (1, 1): qml.PauliY(1),
+            (1, 2): qml.PauliZ(1),
+        })
+
+    Then the generators can be accessed by using the tuples ``(arg_index, par_index)`` as key
+    with ``par_index=None`` for the ``PauliX(0)`` generator. The latter may alternatively
+    be accessed directly via ``generators[1]``.
+    """
+
+    def __init__(self, data=None, single_arg=False, single_object=False, like=None):
+        if like is not None:
+            if not isinstance(like, ArgMap):
+                raise ArgMapError("Trying to inherit properties from non-ArgMap object.")
+            single_arg, single_object = like.single_arg, like.single_object
+        self.single_arg = single_arg
+        self.single_object = single_object
+        _data = self._preprocess_data(data)
+        super().__init__(_data)
+        self.consistency_check()
+
+    def _preprocess_data(self, data):
+        if data is None:
+            data = {}
+        if self.single_object:
+            return {(None, None): data}
+
+        if not isinstance(data, dict):
+            try:
+                data = dict(data)
+            except (ValueError, TypeError) as e:
+                raise ArgMapError(
+                    f"The input could not be interpreted as dictionary; input:\n{data}"
+                ) from e
+
+        return {_interpret_key(key, self.single_arg): val for key, val in data.items()}
```

Comment:
this is relatively minor, but having `dict(data)` above, followed by `for key, val in data.items()` will cause the dictionary to be looped through twice! Probably not too big a deal, unless this function is called repeatedly inside a loop.

Perhaps you could move the try-except loop, and do the loop only once?

Reply:
Oh, this is a good catch, also because I don't see immediately how to avoid this...
I wanted to invoke `dict` to not have to replicate the parsing code within `dict` that allows us to pass, for example, a list of tuples.
I am not quite sure I get the suggestion of moving the try-except loop :thinking: